text/0000-transparent-unions.md

```diff
@@ -0,0 +1,229 @@
+- Feature Name: transparent_enunions
+- Start Date: 2019-02-13
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Allow `#[repr(transparent)]` on `union`s and univariant `enum`s that have exactly one non-zero-sized field (just like `struct`s).
+
+# Motivation
+[motivation]: #motivation
+
+Some `union` types are thin newtype-style wrappers around another type, like `MaybeUninit<T>` (and [once upon a time](https://doc.rust-lang.org/1.28.0/src/core/mem.rs.html#955), `ManuallyDrop<T>`). This type is intended to be used in the same places as `T`, but without being `#[repr(transparent)]` the actual compatibility between it and `T` is left unspecified.
+
+Likewise, some `enum` types only have a single variant, and are similarly thin wrappers around another type.
+
+Making types like these `#[repr(transparent)]` would be useful in certain cases. For example, making the type `Wrapper<T>` (which is a `union` or univariant `enum` with a single field of type `T`) transparent:
+
+- Clearly expresses the intent of the developer.
+- Protects against accidental violations of that intent (e.g., adding a new variant or non-ZST field will result in a compiler error).
+- Makes a clear API guarantee that a `Wrapper<T>` can be transmuted to a `T` or substituted for a `T` in an FFI function's signature.
```

Comment:
> `Wrapper<T>` can be transmuted into a `T` if its layout compatible with `T` - this does not necessarily require `Wrapper<T>` to be `repr(transparent)`.

I didn't say it requires `repr(transparent)` to be transmutable. The primary key is "a clear API guarantee". Two types can be transmutable, but this may not be clear, and may not be an API guarantee.

The caveat about passing uninitialized memory through FFI to external code is a good point. I've added a small note to make that explicit.